### PR TITLE
Private Events, User signup and login, Bootstrap UI

### PR DESCRIPTION
**Pull request by pointerish**
This PR implements the following:

- User creation and login
- Events created by Users
- Users can attend many events
- A user can create many events
- Simple UI built with Bootstrap

### CODE REVIEW

**Comment by pointerish** [comment]
Hi TSEs,

Apologies if the commit history looks weird. I started to work on this using another repo. You can see that one here: [Ensemble](https://github.com/pointerish/Ensemble)
I then migrated to this one due to extreme difficulties with RSpec and Capybara. I did not, mistakenly, start my project with the `-T` option to not add MiniTest and I was not able to set up RSpec/Capybara. It simply did not work.

That's why I pretty much simply copied Models, Controllers, Routes, and Views from the other repo to this one. This project has been a great challenge to me and I learned quite a lot from my struggles and shortcomings. I also learned a lot about myself and I feel like that's the greatest lesson.

Thanks in advance for all the feedback!

**Review by ElijahTobs — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

## Hi @pointerish  :wave: 

You've done well so far :+1:  and I commend you for that :clap: 
You have a close-to-perfect project. Your project is very professional and it stands out from so many other ones I've come across. Kudos. :+1: 
However, there are still some changes to be made before this milestone can be approved.

## REQUIRED

#### Associations

##### User as event creator

* [ ] User model: `has_many :events, foreign_key: :creator_id` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] Event model `belongs_to :creator, class_name: "User"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  * Migrations
    * [ ] events table has creator_id as the foreign key ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

```ruby
def change
  create_table :users do |t|
  end

  create_table :events do |t|
    t.integer :creator_id
  end
end
```

##### User as an Event attendee ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/attendee_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

* [ ] There are 2 equally good options to achieve that, described here https://guides.rubyonrails.org/association_basics.html#choosing-between-has-many-through-and-has-and-belongs-to-many
###### *OPTION 1 - has_and_belongs_to_many*

* [ ] User model `has_and_belongs_to_many :attended_events, class_name: "Event"`
* [ ] Event model `has_and_belongs_to_many :attendees, class_name: "User"`
  * Migrations
    * [ ] *\[OPTION 1\]* manually created joined table

```ruby
def change
  create_table :users do |t|
  end

  create_table :events do |t|
  end

  create_table :users_events do |t|
    t.belongs_to :user, index: true
    t.belongs_to :event, index: true
  end
end
```

```
- [ ] *[OPTION 2]* [create_join_table](https://guides.rubyonrails.org/active_record_migrations.html#creating-a-join-table)
```

```ruby
def change
  create_table :events do |t|
  end

  create_table :users do |t|
  end

  create_join_table :events, :users
end
```

###### *OPTION 2 - has many through*

* [ ] join model is required, its name can be `EventsAttendation` or anything else
* [ ] EventsAttendation model `belongs_to :attendee, class_name: "User"`
* [ ] EventsAttendation model `belongs_to :attended_event, class_name: "Event"`
* [ ] User model `has_many :event_attendations`
* [ ] User model `has_many :attended_events, through: :event_attendations`
* [ ] Event model `has_many :event_attendations`
* [ ] Event model `has_many :attendees, through: :event_attendations`
  * Migrations
    * [ ] join model has corresponding migration

```ruby
def change
  create_table :users do |t|
  end

  create_table :events do |t|
  end

  create_table :event_attendations do |t|
    t.belongs_to :user, index: true
    t.belongs_to :event, index: true
  end
end
```

#### Controllers

##### EventsController

###### `create` action

* [ ] You're expected to use #build method on association, i.e. `current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

#### In your database schema:
* [ ] the events table has column `creator_id` of type `integer`
* [ ] an index for `creator_id` column is added

### Make the requested changes and submit for another review.
## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)

File: db/migrate/20210121170546_add_creator_id_to_events.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddCreatorIdToEvents < ActiveRecord::Migration[6.1]
+  def change
+    add_column :events, :creator_id, :integer
```
Comment (enaburekhan) [inline db/migrate/20210121170546_add_creator_id_to_events.rb:3]: Kindly add the index to the creator_id. The index is used by the rails app to optimize for fast lookup.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Request changes
**Hi Alvarado!**
Great job on making all the changes requested by the previous Tse :clap: :clap:. This project looks almost perfect. However, there's one more change we need to fix in order to get it approved :muscle:.

**Mandatory requirement:**
- [ ] **an index for creator_id column is added. This requirement needs to be met. As I already pointed out below. The indexes are needed to optimize for fast lookup.**

Make the above change and submit another code review request.

You can reach me on slack @ Eric Enaburekhan if you have any questions.
Happy coding :blush:
Eric Enaburekhan

**Review by mahmoud717 — APPROVED** [review]
### STATUS: APPROVED ✔️ ✔️ 

**hi @pointerish**

**Great job with making all of the requirements. the project is approved, you can go ahead and merge this pull request**

**Pull request merged**